File: shared/ios/ReleafCore/Sources/ReleafCoreDrive/DriveClient.swift
/*
 * DriveClient.swift
 * Protocol describing the subset of Google Drive we need.
 *
 * Real implementation: wrap `GoogleAPIClientForREST_Drive`
 * (https://github.com/google/google-api-objectivec-client-for-rest)
 * or call the REST API directly with URLSession.
 */

import Foundation

public struct DriveFile: Equatable, Sendable {
    public let id: String
    public let name: String
    public let mimeType: String
    public let parents: [String]
    public let modifiedTime: Date?

    public init(id: String, name: String, mimeType: String, parents: [String] = [], modifiedTime: Date? = nil) {
        self.id = id
        self.name = name
        self.mimeType = mimeType
        self.parents = parents
        self.modifiedTime = modifiedTime
    }

    public var isFolder: Bool { mimeType == "application/vnd.google-apps.folder" }
}

public enum DriveError: Error, Equatable {
    case unauthenticated
    case notFound
    case notImplemented
    case underlying(String)
}

/// Tiny façade over Drive. Everything is async; auth is passed in per-call
/// so the client doesn't need its own token lifecycle.
public protocol DriveClient: AnyObject, Sendable {
    /// Find or create the top-level user-visible folder (default name: "Inkcreate").
    func ensureRootFolder(named name: String, accessToken: String) async throws -> DriveFile

    /// Find or create a named folder under `parentId`.
    func ensureFolder(named name: String, parentId: String, accessToken: String) async throws -> DriveFile

    /// List direct children of `folderId`.
    func listChildren(of folderId: String, accessToken: String) async throws -> [DriveFile]

    /// Upload / overwrite a small JSON file. `filename` is the Drive file name.
    func uploadJSON(_ data: Data, filename: String, parentId: String, accessToken: String) async throws -> DriveFile

    /// Download a file's bytes.
    func downloadBytes(fileId: String, accessToken: String) async throws -> Data

    /// Move a file to the Drive trash.
    func trash(fileId: String, accessToken: String) async throws
}

/// In-memory stub — lets skeletons + previews render without real Drive calls.
public final class InMemoryDriveClient: DriveClient, @unchecked Sendable {
    private var files: [String: DriveFile] = [:]
    private var blobs: [String: Data] = [:]
    private let queue = DispatchQueue(label: "releaf.InMemoryDriveClient")

    public init() {}

    public func ensureRootFolder(named name: String, accessToken: String) async throws -> DriveFile {
        queue.sync {
            if let existing = files.values.first(where: { $0.name == name && $0.parents.isEmpty && $0.isFolder }) {
                return existing
            }
            let file = DriveFile(id: UUID().uuidString, name: name, mimeType: "application/vnd.google-apps.folder")
            files[file.id] = file
            return file
        }
    }

    public func ensureFolder(named name: String, parentId: String, accessToken: String) async throws -> DriveFile {
        queue.sync {
            if let existing = files.values.first(where: { $0.name == name && $0.parents == [parentId] && $0.isFolder }) {
                return existing
            }
            let file = DriveFile(id: UUID().uuidString, name: name, mimeType: "application/vnd.google-apps.folder", parents: [parentId])
            files[file.id] = file
            return file
        }
    }

    public func listChildren(of folderId: String, accessToken: String) async throws -> [DriveFile] {
        queue.sync { files.values.filter { $0.parents.contains(folderId) } }
    }

    public func uploadJSON(_ data: Data, filename: String, parentId: String, accessToken: String) async throws -> DriveFile {
        queue.sync {
            if let existing = files.values.first(where: { $0.name == filename && $0.parents == [parentId] }) {
                blobs[existing.id] = data
                return existing
            }
            let file = DriveFile(id: UUID().uuidString, name: filename, mimeType: "application/json", parents: [parentId])
            files[file.id] = file
            blobs[file.id] = data
            return file
        }
    }

    public func downloadBytes(fileId: String, accessToken: String) async throws -> Data {
        try queue.sync {
            guard let data = blobs[fileId] else { throw DriveError.notFound }
            return data
        }
    }

    public func trash(fileId: String, accessToken: String) async throws {
        queue.sync {
            files.removeValue(forKey: fileId)
            blobs.removeValue(forKey: fileId)
        }
    }
}

File: shared/android/shared/drive/src/main/kotlin/app/releaf/mobile/data/drive/DriveClient.kt
/*
 * DriveClient.kt
 * Façade over Google Drive. Real implementation wraps the Drive v3 REST API
 * (either google-api-services-drive or plain OkHttp calls).
 */

package app.releaf.mobile.data.drive

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.time.Instant
import java.util.UUID

data class DriveFile(
    val id: String,
    val name: String,
    val mimeType: String,
    val parents: List<String> = emptyList(),
    val modifiedTime: Instant? = null,
) {
    val isFolder: Boolean get() = mimeType == "application/vnd.google-apps.folder"
}

sealed class DriveError(message: String) : Exception(message) {
    object Unauthenticated : DriveError("Unauthenticated")
    object RateLimited     : DriveError("Rate limited")
    object NotFound        : DriveError("Not found")
    object NotImplemented  : DriveError("Not implemented")
    class  Underlying(msg: String) : DriveError(msg)
}

interface DriveClient {
    suspend fun ensureRootFolder(name: String, accessToken: String): DriveFile
    suspend fun ensureFolder(name: String, parentId: String, accessToken: String): DriveFile
    suspend fun listChildren(folderId: String, accessToken: String): List<DriveFile>
    suspend fun uploadJson(data: ByteArray, filename: String, parentId: String, accessToken: String): DriveFile

    /**
     * Upload / overwrite a binary file (PDF, JPEG, etc.) with caller-
     * supplied MIME type. PATCHes an existing same-name child if
     * present so re-uploads don't fork into duplicates.
     */
    suspend fun uploadBinary(
        data: ByteArray,
        filename: String,
        contentType: String,
        parentId: String,
        accessToken: String,
    ): DriveFile
    suspend fun downloadBytes(fileId: String, accessToken: String): ByteArray
    suspend fun trash(fileId: String, accessToken: String)
}

/** In-memory stub — lets skeletons and previews run without network. */
class InMemoryDriveClient : DriveClient {
    private val mutex = Mutex()
    private val files  = mutableMapOf<String, DriveFile>()
    private val blobs  = mutableMapOf<String, ByteArray>()

    override suspend fun ensureRootFolder(name: String, accessToken: String): DriveFile = mutex.withLock {
        files.values.firstOrNull { it.name == name && it.parents.isEmpty() && it.isFolder }
            ?: DriveFile(UUID.randomUUID().toString(), name, "application/vnd.google-apps.folder")
                .also { files[it.id] = it }
    }

    override suspend fun ensureFolder(name: String, parentId: String, accessToken: String): DriveFile = mutex.withLock {
        files.values.firstOrNull { it.name == name && it.parents == listOf(parentId) && it.isFolder }
            ?: DriveFile(UUID.randomUUID().toString(), name, "application/vnd.google-apps.folder", listOf(parentId))
                .also { files[it.id] = it }
    }

    override suspend fun listChildren(folderId: String, accessToken: String): List<DriveFile> = mutex.withLock {
        files.values.filter { folderId in it.parents }
    }

    override suspend fun uploadJson(data: ByteArray, filename: String, parentId: String, accessToken: String): DriveFile =
        uploadBinary(data, filename, "application/json", parentId, accessToken)

    override suspend fun uploadBinary(
        data: ByteArray,
        filename: String,
        contentType: String,
        parentId: String,
        accessToken: String,
    ): DriveFile = mutex.withLock {
        val existing = files.values.firstOrNull { it.name == filename && it.parents == listOf(parentId) }
        if (existing != null) {
            blobs[existing.id] = data
            return@withLock existing
        }
        val file = DriveFile(UUID.randomUUID().toString(), filename, contentType, listOf(parentId))
        files[file.id] = file
        blobs[file.id] = data
        file
    }

    override suspend fun downloadBytes(fileId: String, accessToken: String): ByteArray = mutex.withLock {
        blobs[fileId] ?: throw DriveError.NotFound
    }

    override suspend fun trash(fileId: String, accessToken: String) { mutex.withLock {
        files.remove(fileId); blobs.remove(fileId)
    } }
}
